### PR TITLE
Allow for variables in storyboards

### PR DESCRIPTION
**Pull request by ColdVolcano**

### CODE REVIEW

**Review by smoogipoo — DISMISSED** [review]
👍

File: osu.Game/Beatmaps/Formats/OsuLegacyDecoder.cs [hunk 1]
```diff
@@ -204,6 +208,17 @@ private void handleDifficulty(Beatmap beatmap, string key, string val)
...
+            do
+            {
+                string[] valSplit = val.Split(',');
+                for (int i = 0; i < valSplit.Length; i++)
+                {
+                    if (valSplit[i][0] == '$' && variables.ContainsKey(valSplit[i]))
+                        valSplit[i] = variables[valSplit[i]];
+                }
+                val = string.Join(",", valSplit);
+            } while (val.IndexOf('$') != -1);
```
Comment (peppy) [inline osu.Game/Beatmaps/Formats/OsuLegacyDecoder.cs]: why not while instead of do{}while? also `< 0` not `!= -1`

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Beatmaps/Formats/OsuLegacyDecoder.cs (hunk 1 above)
Comment (smoogipooo) [inline osu.Game/Beatmaps/Formats/OsuLegacyDecoder.cs]: I prefer `!= -1` when it's defined by the spec to be this value.

File: osu.Game/Beatmaps/Formats/OsuLegacyDecoder.cs [hunk 2]
```diff
@@ -391,7 +406,7 @@ protected override void ParseFile(StreamReader stream, Beatmap beatmap)
                 }
 
                 string val = line, key = null;
-                if (section != Section.Events && section != Section.TimingPoints && section != Section.HitObjects)
+                if (section != Section.Events && section != Section.TimingPoints && section != Section.HitObjects && section != Section.Variables)
```
Comment (peppy) [inline osu.Game/Beatmaps/Formats/OsuLegacyDecoder.cs]: this is just getting stupid. can we refactor it?

File: osu.Game/Beatmaps/Formats/OsuLegacyDecoder.cs [hunk 3]
```diff
@@ -426,6 +441,11 @@ protected override void ParseFile(StreamReader stream, Beatmap beatmap)
                             beatmap.HitObjects.Add(obj);
 
                         break;
+                    case Section.Variables:
+                        key = val.Remove(val.IndexOf('=')).Trim();
```
Comment (peppy) [inline osu.Game/Beatmaps/Formats/OsuLegacyDecoder.cs]: `val.Split('=')`?

File: osu.Game/Beatmaps/Formats/OsuLegacyDecoder.cs [hunk 4]
```diff
@@ -204,6 +208,17 @@ private void handleDifficulty(Beatmap beatmap, string key, string val)
 
         private void handleEvents(Beatmap beatmap, string val)
         {
+            do
+            {
+                string[] valSplit = val.Split(',');
+                for (int i = 0; i < valSplit.Length; i++)
+                {
+                    if (valSplit[i][0] == '$' && variables.ContainsKey(valSplit[i]))
```
Comment (peppy) [inline osu.Game/Beatmaps/Formats/OsuLegacyDecoder.cs]: assign valSplit[i] to a local rather than looking it up three times

**Comment by smoogipooo** [comment]
@peppy please re-review/approve.

File: osu.Game/Beatmaps/Formats/OsuLegacyDecoder.cs [hunk 5]
```diff
@@ -56,36 +59,37 @@ private enum Section
             TimingPoints,
             Colours,
             HitObjects,
+            Variables,
         }
 
-        private void handleGeneral(Beatmap beatmap, string key, string val)
```
Comment (peppy) [inline osu.Game/Beatmaps/Formats/OsuLegacyDecoder.cs]: `val` was used here for a reason (generally we avoid using `value` as it is a keyword in certain contexts)

Comment (peppy) [inline osu.Game/Beatmaps/Formats/OsuLegacyDecoder.cs]: even github isn't smart enough to realise it's not a keyword here:

allow_for_variables_in_storyboards_by_coldvolcano_ _pull_request__866_ _ppy_osu

Comment (smoogipooo) [inline osu.Game/Beatmaps/Formats/OsuLegacyDecoder.cs]: Ouch :\

**Review by peppy — APPROVED** [review]

**Pull request merged**